core/src/telemetry/log_export.rs

```diff
@@ -0,0 +1,183 @@
+        let log = CoreLog {
+            target: event.metadata().target().to_string(),
+            // This weird as_str dance prevents adding extra quotes
+            message: message.as_str().unwrap_or_default().to_string(),
+            timestamp: SystemTime::now(),
+            level: *event.metadata().level(),
+            fields,
+            span_contexts: spans,
+        };
+        let _ = self.logs_in.lock().push(log);
+    }
+}
+
+struct JsonVisitor<'a>(&'a mut HashMap<String, serde_json::Value>);
+
+impl<'a> tracing::field::Visit for JsonVisitor<'a> {
+    fn record_f64(&mut self, field: &tracing::field::Field, value: f64) {
+        self.0
+            .insert(field.name().to_string(), serde_json::json!(value));
+    }
+
+    fn record_i64(&mut self, field: &tracing::field::Field, value: i64) {
+        self.0
+            .insert(field.name().to_string(), serde_json::json!(value));
+    }
+
+    fn record_u64(&mut self, field: &tracing::field::Field, value: u64) {
+        self.0
+            .insert(field.name().to_string(), serde_json::json!(value));
+    }
+
+    fn record_bool(&mut self, field: &tracing::field::Field, value: bool) {
+        self.0
+            .insert(field.name().to_string(), serde_json::json!(value));
+    }
+
+    fn record_str(&mut self, field: &tracing::field::Field, value: &str) {
+        self.0
+            .insert(field.name().to_string(), serde_json::json!(value));
+    }
+
+    fn record_error(
+        &mut self,
+        field: &tracing::field::Field,
+        value: &(dyn std::error::Error + 'static),
+    ) {
+        self.0.insert(
+            field.name().to_string(),
+            serde_json::json!(value.to_string()),
+        );
+    }
+
+    fn record_debug(&mut self, field: &tracing::field::Field, value: &dyn std::fmt::Debug) {
+        self.0.insert(
+            field.name().to_string(),
+            serde_json::json!(format!("{:?}", value)),
+        );
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        construct_filter_string, fetch_global_buffered_logs, telemetry_init, Logger,
+        TelemetryOptionsBuilder,
+    };
+    use tracing::Level;
+
+    #[instrument(fields(bros = "brohemian"))]
+    fn instrumented(thing: &str) {
+        warn!("warn");
+        info!(foo = "bar", "info");
+        debug!("debug");
+    }
+
+    #[tokio::test]
+    async fn test_forwarding_output() {
+        let opts = TelemetryOptionsBuilder::default()
+            .logging(Logger::Forward {
+                filter: construct_filter_string(Level::INFO, Level::WARN),
+            })
+            .build()
+            .unwrap();
+        telemetry_init(&opts).unwrap();
+
+        let top_span = span!(Level::INFO, "yayspan", huh = "wat");
+        let _guard = top_span.enter();
+        info!("Whata?");
+        instrumented("hi");
+        info!("Donezo");
+
+        let logs = fetch_global_buffered_logs();
+        for log in logs {
+            println!("{:?}", log);
+        }
```

Comment:
Should this be tested?

Reply:
That is with assertions instead of printing?

Reply:
Yeah, may as well verify a few things at least. Will update.